smb2pdu.c

```diff
@@ -1490,8 +1491,14 @@ static int ntlm_authenticate(struct ksmbd_work *work)
 		    (req->SecurityMode & SMB2_NEGOTIATE_SIGNING_REQUIRED))
 			sess->sign = true;
 
-		if (conn->vals->capabilities & SMB2_GLOBAL_CAP_ENCRYPTION &&
-		    conn->ops->generate_encryptionkey &&
+		/*
+		 * SMB 3.0 and 3.0.2 dialects use the SMB2_GLOBAL_CAP_ENCRYPTION flag.
+		 * SMB 3.1.1 uses the cipher_type field.
+		 */
+		encrequested = conn->cipher_type ||
+			(conn->vals->capabilities & SMB2_GLOBAL_CAP_ENCRYPTION);
```

Comment:
we don't need to use encrequested variable.

Reply:
I created the variable to make clearer the conditions on which the encryption applied, since there were already a sizeable amount of operands in the if statement.

Mixing complex and, ors and bit-twiddling in the same if would be a big no for me since it requires thinking about C's operand priorities, but if you'd prefer I can change it.

Reply:
I've moved the logic into a function, which makes it IMHO clearer and reduces duplicate code.

I'll send it now to the LKML for further review.